### PR TITLE
Use Sparkle sorting/filtering in #livecheck_min_os

### PR DESCRIPTION
**Pull request by samford**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

The `#livecheck_min_os` cask audit is intended to use a Sparkle feed's `minimumSystemVersion` information to ensure that the cask has a correct `depends_on macos: ...` value. It reimplements some of the logic in livecheck's `Sparkle` strategy but naively treats the first `item` in the appcast as newest and doesn't adhere to the `Sparkle` strategy's sorting/filtering.

Some appcasts are sorted from oldest to newest and some have `item`s with out-of-order `pubDate` values, so naively using the first `item` in the audit has led to incorrect audit failures at times. As one example, [this issue affected the `hyperkey` cask](https://github.com/Homebrew/homebrew-cask/pull/136777) until upstream updated the appcast to sort items from newest to oldest. This issue is also mitigated by #16013, as that PR introduced logic to check the minimum system version from cask artifact plist files in addition to the existing `livecheck` block audit.

To be able to easily use the `Sparkle` strategy's `Item` sorting/filtering logic in the audit, I've refactored that code into `#filter_items` and `#sort_items` methods. As a side effect, this also means that the `#items_from_content` method doesn't filter (or sort) items anymore, which may be useful in the future (e.g., if we ever need to add the ability to pass unfiltered/unsorted items to a `strategy` block).

That said, this PR reimplements `#livecheck_min_os` to use `Livecheck::Strategy#page_content` to fetch the appcast content, `#items_from_content` to parse the appcast, and the new `#filter_items` and `#sort_items` methods. This aligns the audit with the `Sparkle` strategy's behavior and resolves the aforementioned issue.

-----

If you wish to manually replicate this issue (for testing), you can edit the `hyperkey` cask to use an old snapshot of the appcast and the `version`/`depends_on` values from the previously-mentioned cask PR:

```diff
diff --git a/Casks/h/hyperkey.rb b/Casks/h/hyperkey.rb
index 59a73ad96678d1b8a70b6aea5cee151d4f5ff21a..c8766630af3cc3d5f04cbbbc70c703d70da6c1ec 100644
--- a/Casks/h/hyperkey.rb
+++ b/Casks/h/hyperkey.rb
@@ -1,6 +1,6 @@
 cask "hyperkey" do
-  version "0.23"
-  sha256 "8e8435b734737e0f08dc1278a2221e9f1a20ee75422a253b25503669a47fe9a1"
+  version "0.22"
+  sha256 "b4b78d923bb38424ceb302d595713b4541e0dd119e40aadb1e42d04d53d731b6"
 
   url "https://hyperkey.app/downloads/Hyperkey#{version}.dmg"
   name "Hyperkey"
@@ -8,12 +8,12 @@ cask "hyperkey" do
   homepage "https://hyperkey.app/"
 
   livecheck do
-    url "https://hyperkey.app/downloads/updates.xml"
+    url "https://web.archive.org/web/20230618095239id_/https://hyperkey.app/downloads/updates.xml"
     strategy :sparkle, &:short_version
   end
 
   auto_updates true
-  depends_on macos: ">= :catalina"
+  depends_on macos: ">= :high_sierra"
 
   app "Hyperkey.app"
 

```

On the [brew] `master` branch, you will also need to set `plist_min_os` to `nil` in `#audit_min_os` to ensure that only the `#livecheck_min_os` value is used. This is the simplest way of replicating the behavior before #16013 introduced a mitigation.

```diff
diff --git a/Library/Homebrew/cask/audit.rb b/Library/Homebrew/cask/audit.rb
index a01a8b136b8796ec1a5041dc8f751fcf6b0899ee..4cfceaf89715b23d2a6f6bb63789965bac303635 100644
--- a/Library/Homebrew/cask/audit.rb
+++ b/Library/Homebrew/cask/audit.rb
@@ -567,7 +567,8 @@ module Cask
 
       odebug "Auditing minimum OS version"
 
-      plist_min_os = cask_plist_min_os
+      # plist_min_os = cask_plist_min_os
+      plist_min_os = nil
       sparkle_min_os = livecheck_min_os
 
       debug_messages = []

```

With those changes in place, `brew audit --cask --online --strict hyperkey` should fail with `Upstream defined :sierra as the minimum OS version and the cask defined :high_sierra` (basically the same as the previous `hyperkey` PR). [Fair warning, archive.org is being weird and you may need to use `20230618095239` or `20230618095239im_` in the URL to get the unmodified XML (despite both being incorrect). `id_` is the correct way of fetching the unmodified file but the server alternates between returning it from only one of these URLs at a time (probably due to a caching bug 🤷).]

After that, you can check out this PR branch (keeping the `plist_min_os = nil` change in place) and the same `brew audit` command should correctly pass.

-----

This PR also modifies `#items_from_content` to use a `MacOSVersion` object as the `Item.minimum_system_version` value instead of a string. This reduces boilerplate when making comparisons (e.g., `item.minimum_system_version >= :ventura`) in `Sparkle#filter_items`, `strategy` blocks, and the `#livecheck_min_os` audit method, by removing the need to replicate the `MacOSVersion` error handling everywhere. I wanted to include this as part of #16158 but it was easier to do as part of the aforementioned refactoring.

Somewhat related, I've also added `#presence` calls to `Item` value assignments, to ensure that empty strings become `nil`. While debugging, I was surprised to find that some `Item` values were an empty string instead of `nil`, so this ensures that the values are consistent/predictable.

-----

Lastly, this modifies `Cask::Audit#livecheck_min_os` to skip `livecheck` blocks with a `Sparkle` `strategy` block using the `items` argument (instead of `item`), as these ignore/override the `Sparkle` strategy's default sorting such that we can't predictably replicate the logic in the audit method.

To be clear, we could execute the `strategy` block but it will only return version strings. The source of the version information varies between `strategy` blocks, so it's not really feasible to correlate versions to `Item` objects to identify the `minimumSystemVersion` of the newest version. However, `#cask_plist_min_os` may take care of those instances, so it's not a total loss.

### CODE REVIEW

File: Library/Homebrew/cask/audit.rb [hunk 1]
```diff
@@ -597,31 +597,29 @@ def livecheck_min_os
       return unless cask.livecheckable?
       return if cask.livecheck.strategy != :sparkle
 
-      out, _, status = curl_output("--fail", "--silent", "--location", cask.livecheck.url)
-      return unless status.success?
+      # `Sparkle` strategy blocks that use the `items` argument (instead of
+      # `item`) contain arbitrary logic that ignores/overrides the strategy's
+      # sorting, so we can't identify which item would be first/newest here.
+      return if cask.livecheck.strategy_block.present? &&
+                cask.livecheck.strategy_block.parameters[0] == [:opt, :items]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:604]: This seems to know way too much about the internals of this strategy block and would be better placed with something like a `cannot_do_this?` method.

File: Library/Homebrew/cask/audit.rb [hunk 2]
```diff
@@ -597,31 +597,29 @@ def livecheck_min_os
...
+            Homebrew::Livecheck::Strategy::Sparkle.items_from_content(content),
+          ),
+        )
+      rescue
+        return
       end
+      return if items.blank?
+
+      min_os = items[0]&.minimum_system_version&.strip_patch
+      min_os = MacOSVersion.new("11") if min_os == "10.16"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/audit.rb:621]: I know it's not your code but rereading as it is moved around: a comment here might be nice explaining this.

File: Library/Homebrew/livecheck/strategy/sparkle.rb [hunk 3]
```diff
@@ -86,19 +86,29 @@ def self.items_from_content(content)
...
-              os = enclosure["os"]
+              url = enclosure["url"].presence
+              short_version = enclosure["shortVersionString"].presence
+              version = enclosure["version"].presence
+              os = enclosure["os"].presence
             end
 
-            title = item.elements["title"]&.text&.strip
-            link = item.elements["link"]&.text&.strip
+            title = item.elements["title"]&.text&.strip&.presence
```
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/sparkle.rb:95]: That's a lot of `&`, a refactoring might be nice at some point given these are all doing the same thing (`&.text&.strip&.presence`)

File: Library/Homebrew/livecheck/strategy/sparkle.rb [hunk 4]
```diff
@@ -146,6 +144,33 @@ def self.items_from_content(content)
...
 
+        # Filters out items that aren't suitable for Homebrew.
+        #
+        # @param items [Array] appcast items
+        # @return [Array]
+        sig { params(items: T::Array[Item]).returns(T::Array[Item]) }
+        def self.filter_items(items)
+          items.select do |item|
+            # Omit non-macOS items
+            next false if item.os && !((item.os == "osx") || (item.os == "macos"))
```
Comment (MikeMcQuaid) [inline Library/Homebrew/livecheck/strategy/sparkle.rb:155]: `!((item.os == "osx") || (item.os == "macos"))` is a bit hard to parse, would be nice to split it out into variables or similar

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks @samford, almost there, a few nits.

File: Library/Homebrew/cask/audit.rb (hunk 1 above)
Comment (samford) [reply Library/Homebrew/cask/audit.rb:604]: I'm not sure that I follow. Could you explain a bit more?

For context, `Sparkle` `strategy` blocks that use the `items` argument will categorically do something instead of or in addition to the default sorting. The usual argument is `item` (i.e., the first item after sorting/filtering) and we only use `items` when there's some issue where the first-sorted item isn't the newest version (e.g., an older version may have a more recent `pubDate`).

File: Library/Homebrew/cask/audit.rb (hunk 2 above)
Comment (samford) [reply Library/Homebrew/cask/audit.rb:621]: I'll add a comment with the next push 👍

File: Library/Homebrew/livecheck/strategy/sparkle.rb (hunk 3 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:95]: I'd been kind of thinking the same thing, even before adding the `#presence` calls. We could add a method for this in the `Xml` strategy:

```ruby
def self.element_text(parent, el_name)
  parent.get_elements(el_name)&.first&.text&.strip&.presence
end
```

The `#elements` method we're currently using works but it doesn't exist in the `REXML` docs, so this would also allow us to switch to `get_elements(...)&.first` without worrying about the added verbosity. [This switch would be necessary to get `brew typecheck` to pass anyway.]

Using this method, this code would end up like: `Xml.element_text(item, "title")`.

Anyway, this is just to say that I agree with the idea, it's an easy addition, and I can take care of it in a follow-up PR 👍

File: Library/Homebrew/livecheck/strategy/sparkle.rb (hunk 4 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:155]: If we add an array constant for this (`MACOS_STRINGS = ["macos", "osx"].freeze`), we could use `if item.os && MACOS_STRINGS.none?(item.os)` instead. Is that more understandable or did you have something else in mind?

File: Library/Homebrew/livecheck/strategy/sparkle.rb (hunk 3 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:95]: > Using this method, this code would end up like: `Xml.element_text(item, "title")`.

This seems nicer.

> We could add a method for this in the `Xml` strategy

I'm not sure all the `&` are necessary here: better to do early `return` IMO in a method like that on the specific cases that may be `nil` rather than having to chain all of them. It's more code but a bit more explicit what's going on.

File: Library/Homebrew/livecheck/strategy/sparkle.rb (hunk 4 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:155]: They're common values in [Sparkle appcasts](https://sparkle-project.org/documentation/publishing/#alternate-download-locations-for-other-operating-systems). Since this is already in a `Sparkle` class (and `SPARKLE_` would feel repetitive), how about `APPCAST_MACOS_STRINGS`?

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:155]: That's fine, just a comment above the variable explaining that would be handy!

File: Library/Homebrew/livecheck/strategy/sparkle.rb (hunk 3 above)
Comment (samford) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:95]: > I'm not sure all the `&` are necessary here: better to do early `return`

I agree (the method above was just a rough idea). Using early `return`s, it looks something like this:

```ruby
def self.element_text(parent, el_name)
  element = parent.get_elements(el_name).first
  return if element.nil?

  text = element.text
  return if text.blank?

  text.strip
end
```

The first condition could also be `unless element`, `if element.blank?`, etc. but we can work out the fine details in a follow-up PR.

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/sparkle.rb:95]: Looks lovely 😍

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @samford!
